### PR TITLE
Refactor/must get agent activity logic

### LINKED ISSUE #5308: Review filter_then_check for any potential improvements
Follow up to https://github.com/holochain/holochain/issues/5278

Part of performance improvements to `must_get_agent_activity` requests

The function `filter_then_check` is taking a lot of the time of the overall function all. Review it to see if performance improvements from refactoring are feasible, and if so implement them.

https://github.com/holochain/holochain/blob/90b6b2399ee0673962fb3cb4a8e4a546b01b7c5e/crates/holochain_types/src/chain.rs#L477

### PR DESCRIPTION
**Pull request by mattyg**
### Summary
Resolves #5308 

Refactor of must_get_agent_activity logic for correctness. Performance with chain lengths up to 5K appears roughly the same as the current implementation.

This modifies the ChainFilter, removing the ability to specify multiple filters or hashes. Now only a single take, until hash, or until timestamp can be specified. See the changelog for the exact behavior.

### TODO:
- [x] CHANGELOGs updated with appropriate info
- [x] All code changes are reflected in docs, including module-level docs

## Release Notes

* **Breaking Changes**
  * `ChainFilter` API refactored: replaced builder pattern with direct constructor methods
  * `MustGetAgentActivityResponse` variants updated: `EmptyRange` removed; added `UntilHashMissing` and `UntilTimestampIndeterminate` for clearer error handling

* **Improvements**
  * Enhanced fork detection: forked chain branches are now properly excluded
  * Better timestamp determinism: indeterminate cases now clearly reported
  * Improved error messages for invalid filter conditions

### CODE REVIEW

File: crates/holochain_types/src/chain.rs
Comment (mattyg) [inline crates/holochain_types/src/chain.rs:1]: Previously, this contained most of the logic for walking the chain, excluding forks and checking completeness.

File: crates/holochain_cascade/src/lib.rs
Comment (mattyg) [inline crates/holochain_cascade/src/lib.rs:1035]: This is the top-level logic for must_get_agent_activity. Most of the logic is split up into smaller functions located in `./authority/get_agent_activity_query/must_get_agent_activity.rs`

File: crates/holochain_integrity_types/src/chain.rs [hunk 1]
```diff
@@ -17,11 +16,9 @@ mod test;
...
-/// Multiple [`ActionHash`]es can be provided. The filter will stop at and take the first one found.
-/// When providing a Timestamp, the filter will stop at and take the oldest action that has a
-/// Timestamp newer than the provided one. In the case of multiple actions having the same
-/// Timestamp as the limit condition, the filter will stop at the action with the lowest sequence.
-/// Multiple limit conditions can be set. Whichever is the smaller set will be kept.
+/// When providing a Timestamp, the filter returns all actions with timestamp greater than or equal
+/// to the provided one. If multiple actions have a timestamp exactly equal to the limit
+/// timestamp, all of those actions are included.
 #[derive(Serialize, Deserialize, SerializedBytes, Debug, PartialEq, Eq, Hash, Clone)]
 pub struct ChainFilter<H: Eq + Ord + std::hash::Hash = ActionHash> {
```
Comment (mattyg) [inline crates/holochain_integrity_types/src/chain.rs:19]: ChainFilter's LimitConditions no longer supports multiple filter conditions:

- UntilHash takes a single hash
- Multiple has been removed

The builder fns have been removed in favor of a constructor for each filter type to reflect this.

File: crates/holochain_types/src/chain/test.rs
Comment (mattyg) [inline crates/holochain_types/src/chain/test.rs:1]: These have all been moved into holochain_cacade tests

File: crates/holochain_cascade/tests/tests/must_get_agent_activity.rs
Comment (mattyg) [inline crates/holochain_cascade/tests/tests/must_get_agent_activity.rs:1]: Existing tests moved here. More tests added to cover changes to chain filter logic and missing edge cases.

File: crates/holochain_cascade/tests/tests/get_agent_activity.rs
Comment (mattyg) [inline crates/holochain_cascade/tests/tests/get_agent_activity.rs:1]: Renamed and moved must_get_agent_activity tests into their own file

File: crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity/test.rs
Comment (mattyg) [inline crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity/test.rs:1]: Cleaned up existing tests, added more for new logic and edge cases

File: crates/holochain_cascade/src/lib.rs [hunk 2]
```diff
@@ -518,29 +531,64 @@ impl CascadeImpl {
...
+                return Err(e.into());
             }
             Err(e) => {
                 self.record_fetch_error("must_get_agent_activity");
                 return Err(e.into());
             }
         };
 
-        self.add_activity_into_cache(results).await
+        // TODO: Handle conflict and/or merge the responses.
```
Comment (mattyg) [inline crates/holochain_cascade/src/lib.rs]: This is the same logic, but was moved into the function to keep it DRY

File: crates/holochain_types/src/chain.rs [hunk 3]
```diff
@@ -106,12 +21,21 @@ pub enum MustGetAgentActivityResponse {
...
     IncompleteChain,
+    /// The requested until hash was not found in the retained chain range.
+    UntilHashMissing(ActionHash),
+    /// The requested until timestamp range could not be proven complete.
+    ///
+    /// This includes the case where no returned actions satisfy the timestamp,
+    /// and the case where actions satisfy it but there is no deterministic
+    /// lower-bound witness (an action with timestamp below the limit) and the
+    /// returned chain does not reach genesis.
+    UntilTimestampIndeterminate(Timestamp),
```
Comment (mattyg) [inline crates/holochain_types/src/chain.rs:37]: Split out additional response types to clarify when we cannot be sure if we have the complete set of the filtered activity, because we don't have the terminus of the filtered activity.

File: crates/holochain_cascade/src/lib.rs [hunk 4]
```diff
@@ -978,83 +1024,218 @@ impl CascadeImpl {
         Ok(links.len())
     }
 
-    /// Request a hash bounded chain query.
+    /// Look up an action's sequence and timestamp in local stores.
+    ///
+    /// Checks scratch first, then falls back to DHT databases.
+    /// Returns the first match found, or `None` if the action is not in any store.
+    async fn find_action_in_stores(
```
Comment (mattyg) [inline crates/holochain_cascade/src/lib.rs:1002]: This was pulled out to of `must_get_agent_activity` to keep it DRY.

File: crates/holochain_cascade/src/lib.rs [hunk 5]
```diff
@@ -1116,28 +1297,8 @@ impl CascadeImpl {
...
                 .fetch_agent_activity(agent.clone(), query.clone(), options.clone())
                 .await?;
-
             let merged_response: AgentActivityResponse =
                 agent_activity::merge_activities(agent.clone(), &options, results)?;
-
-            // If there is a scratch and warrants were returned, add them to the scratch.
-            // Only warrants coming from the network should be added to the scratch. Locally
-            // found warrants shouldn't be redundantly added to the database.
-            if !authority && !merged_response.warrants.is_empty() {
```
Comment (mattyg) [inline crates/holochain_cascade/src/lib.rs]: This was moved into the `fetch_agent_activity` fn above to keep it DRY.

File: crates/holochain_cascade/tests/tests/get_agent_activity.rs [hunk 6]
```diff
@@ -456,30 +446,6 @@ async fn get_activity_with_warrants() {
     assert_eq!(r2, expected);
 }
 
-#[derive(Default)]
-struct Data {
```
Comment (mattyg) [inline crates/holochain_cascade/tests/tests/get_agent_activity.rs:460]: These tests and test helpers have been moved.

File: crates/holochain_types/src/chain.rs [hunk 7]
```diff
@@ -124,647 +60,3 @@ impl MustGetAgentActivityResponse {
         }
     }
 }
-
-/// Identical structure to [`MustGetAgentActivityResponse`] except it includes
-/// the [`ChainFilterRange`] that was used to produce the response. Doesn't need
-/// to be serialized because it is only used internally.
-/// Note that this is not the final return value of `must_get_agent_activity`.
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub enum BoundedMustGetAgentActivityResponse {
```
Comment (mattyg) [inline crates/holochain_types/src/chain.rs:133]: Previously this contained the core logic for filtering the activity, excluding forks, checking completeness.

File: crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity/test.rs [hunk 8]
```diff
@@ -131,33 +123,326 @@ async fn returns_full_sequence_from_filter(
...
+        create_activity_with_prev(1, hash1_fork, hash0.clone()), // Fork at seq 1
+        create_activity_with_prev(0, hash0, action_hash(&[9])),
+    ];
+    exclude_forked_activity(&mut activity, &action_hash(&[3]));
+
+    assert_eq!(activity.len(), 4);
+
+    let seqs: Vec<u32> = activity.iter().map(|a| a.action.seq()).collect();
+
+    assert_eq!(seqs, vec![3, 2, 1, 0]);
```
Comment (jost-s) [inline crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity/test.rs:252]: I think this should assert hashes as well.

**Review by jost-s — APPROVED** [review]
Looks good to me! To a large extent I rely on the test coverage.

File: crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity/test.rs (hunk 8 above)
Comment (ThetaSinner) [reply crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity/test.rs:252]: Agreed, added that change to multiple tests in this file
